[자동차 경주 게임] 철시 미션 제출합니다.

pkch93 opened the pull request:
오랜만에 자동차 경주 게임 코드를 짜보니깐 재밌네요 ㅎㅎ

미션 짤 시간이 이번주에는 오늘 밖에 안될 것 같아 지금 미션 제출합니다.

잘 부탁드려요 코니~!

pkch93 committed:
setting: gitignore

pkch93 committed:
project settings

pkch93 committed:
docs: README.md

요구사항 정리

pkch93 committed:
feat: Car

- 이동 구현
- 이름 제약사항 구현

pkch93 committed:
feat: Cars

pkch93 committed:
refactor: 접근 제어자 변경

pkch93 committed:
refactor: 매직넘버 상수화

pkch93 committed:
feat: RacingGame.generateGameResult

최종 주행 결과를 도출하는 generateGameResult 메서드 구현

- 이동한 차를 표현한 클래스 MovedCar 구현
- 최종 주형 결과를 담는 GameResult 클래스 구현

pkch93 committed:
feat: Cars.round

한번의 주행 결과를 도출하는 round 함수 구현

- 해당 주행 결과를 담고있는 RoundResult 클래스 구현

pkch93 committed:
feat: RacingGame.startGame

라운드 횟수를 받아 게임을 시작하는 startGame 구현

- 모든 라운드의 결과를 담고 있는 RoundResults 구현
- cars의 갯수를 파악하기 위해 Cars.count 구현
- 난수를 생성해주는 MovableGenerator.generate 구현

pkch93 committed:
feat: GameResult.calculateGameWinner

게임의 최종 주행 결과를 통해 우승자를 도출하는 calculateGameWinner 구현

- 최종 우승자를 관리하는 클래스 Winner 구현
- 우승자 계산을 위해 MovedCar.isWinner 메서드 구현
- 우승자 계산시 계산할 MovedCar가 없다면 NotExistMovedCarException 발생

pkch93 committed:
refactor: RacingGameTest의 변수들 접근제어자 default -> private

pkch93 committed:
refactor: count -> numberOfCars

pkch93 committed:
feat: 입력 및 출력 기능 구현

pkch93 committed:
refactor: NotExistMovedCarException 에러 메세지 변경

pkch93 committed:
refactor: 접근제어자 변경 및 Exception 클래스 exception 패키지로 이동

choihz commented on src/main/java/racingcar/InputView.java:
```diff
@@ -0,0 +1,7 @@
+package racingcar;
```
`InputView`와 `OutputView` 인터페이스가 view 패키지 밖에 있는 이유가 있을까요?

pkch93 replied:
현재는 InputView랑 OutputView를 패키지 안에 둬도 문제는 없겠네요 :)

제가 인터페이스를 패키지 밖으로 뺀 이유는 패키지 간 순환구조 문젝가 있을까봐 InputView랑 OutputView를 사용하고 있는 바깥 패키지에 함께 뒀었어요.

[순환 의존성 문제 참고](http://toby.epril.com/?p=263)

패키지 순환 참조 관련 더 좋은 글을 찾아보려고 했는데 잘 없네요;; 위 글에 순환 구조에 대해 언급하고 있으니 참고해보셔도 좋을 것 같습니다 !

저런 패키지 간 순환 구조를 끊는 방법 중 하나가 DIP(의존성 역전 원칙)을 적용하는 걸로 알고 있어요. 따라서 추상화 대상이 되는 InputView와 OutputView를 사용하고 있는 Main과 InputInterator와 같은 레벨로 두었습니다...

어려운 문제인거 같아요.. 제가 InputView랑 OutputView를 패키지 밖에 둔 이유는 위와 같구요 궁금하신점이나 의논해보면 좋을 부분이 있다면 같이 토의해봐도 좋을것 같습니다 !

choihz replied:
역시....👍 리뷰를 달았는데 도리어 제게 도움이 되었군요. 철시 최고예요! 저도 관련해서 조금 더 공부해 볼게요.

choihz commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,38 @@
+public class Car {
+    private static final int MIN_MOVABLE_NUMBER = 4;
+    private static final int MOVE = 1;
+    private static final int STOP = 0;
```
훌륭합니다 👍

choihz commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,37 @@
+        return cars.get(index);
+    }
+
+    public int numberOfCars() {
```
메서드의 네임 컨벤션은 동사형이었죠!

pkch93 replied:
감사합니다 :)

choihz commented on src/main/java/racingcar/view/ConsoleInputView.java:
```diff
@@ -0,0 +1,32 @@
+    public String carNames() {
+        String names;
+        do {
+            System.out.println("경주할 자동차 이름을 입력하세요.(이름은 쉼표(,) 기준으로 구분");
```
문자열 뒤에 닫는 괄호가 빠졌네요 :)

choihz commented on src/main/java/racingcar/view/ConsoleInputView.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.Scanner;
+
+public class ConsoleInputView implements InputView {
```
`ConsoleInputView` 훌륭합니다 👍

choihz commented on src/main/java/racingcar/view/InputValidator.java:
```diff
@@ -0,0 +1,30 @@
+    private static final Pattern namesRegex = Pattern.compile("([a-zA-Z가-힣]{1,5})(,[a-zA-Z가-힣]{1,5})*");
+
+    static boolean validName(String names) {
+        return !namesRegex.matcher(names).matches();
```
지금 게임 횟수를 콘솔에서 잘못 입력했을 때는 에러 메시지가 잘 떠서 왜 틀렸는지 확인할 수 있는데요, 이름을 잘못 입력했을 때는 관련 메시지가 뜨지 않아 사용자가 왜 틀렸는지 모를 수도 있을 것 같아요. 여기서도 에러 메시지를 통해 사용자에게 정보를 주면 어떨까요?

pkch93 replied:
반영했습니다 :)

choihz commented on src/main/java/racingcar/view/InputValidator.java:
```diff
@@ -0,0 +1,30 @@
+            validNegativeNumber(rounds);
+            return false;
+        } catch (NumberFormatException | NegativeRoundsException e) {
+            System.out.println("횟수는 0 이상의 숫자만 입력가능합니다.");
```
`System.err.println()`도 활용해보세요! 😉

choihz replied:
이 메서드나 위의 메서드나, 옳은 입력값에 대해 false가 리턴되는군요(do-while문을 돌려야 하니까요). 그렇다면 메서드의 이름이 바뀌어야 더 알아보기 쉬울 것 같아요. 그리고 메서드 네이밍 컨벤션은 동사형!

pkch93 replied:
좋은 피드백 감사합니다 :)

choihz commented on src/main/java/racingcar/InputInteractor.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.stream.Stream;
+
+public class InputInteractor {
+    private static final String CARNAME_DELIMITER = ",";
```
`CAR_NAME_DELIMITER`가 더 알아보기 좋겠죠?

pkch93 replied:
왜 CAR 뒤엔 `_`를 안붙였을까요? ㅜㅠㅠ 감사합니다 :)

choihz commented on src/main/java/racingcar/view/InputValidator.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    private static void validNegativeNumber(int rounds) {
+        if (rounds < 0) {
```
0도 포함해야 하지 않을까요? 지금은 0을 입력해도 게임이 실행돼요!

pkch93 replied:
좋은 피드백 감사합니다 :) 1을 MIN_VALID_ROUND로 상수화해서 반영했습니다 !

choihz reviewed (changes requested):
철시~ 제일 빠르면서도 역시 잘하시는군요 👏
특히 스트림을 아주 자유자재로 다루시네요. 보고 정말 많이 배웠어요! 저도 바로 리팩토링 하러 갑니다...
코멘트는 참고하시고 언제든 의견 주세요. 고생 많으셨습니다~

pkch93 committed:
refactor: CARNAME_DELIMITER -> CAR_NAME_DELIMITER

pkch93 committed:
refactor: inputView 메서드 명 변경

- 메서드명을 동사로 드러내도록 input*로 변경
- 시작 메세지에 누락된 닫는 괄호 추가

pkch93 committed:
refactor: Cars.numberOfCars -> Cars.countOfCars

pkch93 committed:
refactor: InputValidator

- validName, validNumberOfRound가 부적절한 경우를 찾으므로 invalid*로 변경
- invalidName에서 부적절한 입력의 경우 오류 메세지 출력
- ConsoleInputView에서 이름 입력 메세지 한번만 출력되도록 변경

choihz commented:
이만 머지하도록 할게요. 고생하셨습니다!
